packages/modelence/src/data/store.ts

```diff
@@ -109,6 +128,86 @@ export class Store<
     return serializeModelSchema(this.schema);
   }
 
+  /**
+   * Extends the store with additional schema fields, indexes, methods, and search indexes.
+   * Returns a new Store instance with the extended schema and updated types.
+   * Methods from the original store are preserved with updated type signatures.
+   *
+   * @param config - Additional schema fields, indexes, methods, and search indexes to add
+   * @returns A new Store instance with the extended schema
+   *
+   * @example
+   * ```ts
+   * // Extend the users collection
+   * export const dbUsers = baseUsersCollection.extend({
+   *   schema: {
+   *     firstName: schema.string(),
+   *     lastName: schema.string(),
+   *     companyId: schema.objectId().optional(),
+   *   },
+   *   indexes: [
+   *     { key: { companyId: 1 } },
+   *     { key: { lastName: 1, firstName: 1 } },
+   *   ],
+   *   methods: {
+   *     getFullName() {
+   *       return `${this.firstName} ${this.lastName}`;
+   *     }
+   *   }
+   * });
+   *
+   * // Now fully typed with new fields
+   * const user = await dbUsers.findOne({ firstName: 'John' });
+   * console.log(user?.getFullName());
+   * ```
+   */
+  extend<
+    TExtendedSchema extends ModelSchema,
+    TExtendedMethods extends Record<
+      string,
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      (this: WithId<InferDocumentType<TSchema & TExtendedSchema>> & any, ...args: any[]) => any
+    > = Record<string, never>,
+  >(config: {
+    schema?: TExtendedSchema;
+    indexes?: IndexDescription[];
+    methods?: TExtendedMethods;
+    searchIndexes?: SearchIndexDescription[];
+  }): Store<
+    TSchema & TExtendedSchema,
+    PreserveMethodsForExtendedSchema<TMethods, TSchema & TExtendedSchema> & TExtendedMethods
+  > {
+    const extendedSchema = {
+      ...this.schema,
+      ...(config.schema || {}),
+    } as TSchema & TExtendedSchema;
+
+    const extendedIndexes = [...this.indexes, ...(config.indexes || [])];
+    const extendedSearchIndexes = [...this.searchIndexes, ...(config.searchIndexes || [])];
+
+    type CombinedMethods = PreserveMethodsForExtendedSchema<TMethods, TSchema & TExtendedSchema> &
+      TExtendedMethods;
+
+    const combinedMethods = {
+      ...(this.methods || {}),
+      ...(config.methods || {}),
+    } as CombinedMethods | undefined;
+
+    const extendedStore = new Store<TSchema & TExtendedSchema, CombinedMethods>(this.name, {
+      schema: extendedSchema,
+      methods: combinedMethods as unknown as CombinedMethods | undefined,
+      indexes: extendedIndexes,
+      searchIndexes: extendedSearchIndexes,
+    });
+
+    // If this store is already initialized, initialize the extended store too
+    if (this.client) {
+      extendedStore.init(this.client);
+    }
```

Comment:
@omegascorp shouldn't we just throw an error if the store is already initialized? We create indexes on startup, and there is a specific time & place for it when the app starts, so if you've missed that we probably shouldn't allow modifying anything after that, since that means the user is doing something wrong. If you extend an existing Store after things have been initialized, it means you are already creating something too late, after `startApp` has been called, and your extended version won't be properly provisioned and its indexes won't be created.